---
title: "Flexible iterative fitting over omics-type data using the modeling algorithm of your choice"
output: rmarkdown::html_vignette
bibliography: references.bib
author: "Daniel HammarstrÃ¶m"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Flexible iterative fitting over omics-type data using the modelling algorithm of your choice}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



## Background

Differential expression is determined in RNA-sequencing data using specific software packages such as edgeR [@McCarthy2012] or DEseq2 [@Love2014]. These packages provides efficient algorithms for fitting gene-expression data to count-based models by sharing information across genes in relatively simple designs. More elaborate experimental designs may require the addition of random effects to such models when observations are not independent. To accommodate random effects Vestal et al. [@Vestal2022] developed [lmerSeq](https://github.com/stop-pre16/lmerSeq) which allowed iterative fitting of mixed linear models to transformed count data. Similarly, Lewis et al. [@Lewis] developed [glmmSeq](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html#glmmSeq) to fit negative binomial mixed effects models. We have previously used negative binomial mixed models to fit data from repeated measures designs [@Khan2020; @Mlmen2021], however the software implementation was not formalized to user-friendly functions.

The aim of the `seqwrap` package is to provide even more flexibility to model fitting of omics-type data. Although [lmerSeq](https://github.com/stop-pre16/lmerSeq) and [glmmSeq](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html#glmmSeq) both provides powerful implementations of fitting algorithms to iterative fitting they are limited as the user cannot select which fitting algorithm to use. `seqwrap` can use any available fitting algorithm (`stats::lm`, `glmmTMB::glmmTMB`, `lme4::lmer`, etc.) which makes the model fitting process more flexible. In addition, the user can provide user-created summary- and model evaluation-functions to tailor output for specific purposes.

## Iterative fitting of high dimensional data

The `seqwrap` function lets the user specify which modelling algorithm to be used in modelling data provided as a data frame of counts (or similar) and metadata. Target-specific models are passed to the results together with optional user-specified summary and model-diagnostic functions. The function aims to not limit the user in preferences of specific models. This allows for comparisons between modelling frameworks and flexibility for many types of omics-type data. The basic outline of the function can be seen in the diagram below.


```{r diagram}
#| echo: false
#| fig-cap: "Basic structure of the `seq_wrapper` function.
#| Data and metadata together with user-selected modelling
#| algorithms and arguments are combined in parallel to fit
#| model objects. Results are returned containing models and
#| any results from user-constructed summaries or diagnostics."

library(DiagrammeR)

DiagrammeR::grViz(" # All instructions are within a large character string
digraph surveillance_diagram {
# 'digraph' means 'directional graph', then the graph name

  # graph statement
  #################
  forcelabels=true;
  graph [layout = dot,
         rankdir = TB,
         overlap = false,
         fontsize = 8]


  # nodes (circles)
  #################
  node [shape = circle,                  # shape = circle
       fixedsize = true
       width = 1.3]                      # width of circles


    data   [label = 'High-dimension\ndata']
    metadata [label = 'Meta data']
    model [label = 'Modelling\nalgorithm']
    modelargs [label = 'Model\narguments']

    fit [label = 'Model\nobjects']
    summaries [label = 'Model\nsummaries']
    evaluations [label = 'Model\ndiagnostics']

    output [label = 'Results']


  # edges
  #######

  data -> fit
  metadata -> fit
  model -> fit
  modelargs -> fit

  fit -> summaries
  fit -> evaluations

  fit -> output
  summaries -> output
  evaluations -> output


}
")
```


### Fitting count data to a negative binomial model using `glmmTBM`


#### Required packages and sample data

The data set used in this demonstration has 1000 rows with each row representing a target transcript. The first column in the data set has the transcript identifier and remaining columns represents samples (n = 196) from 25 individuals. The design of the study can be seen below in @fig-design.

```{r}
#| label: packages
library(tidyverse)
library(glmmTMB)
library(parallel)

# load sample data
data(rna_seq_sample)
data(rna_seq_metadata)
```


```{r}
#| label: fig-design
#| fig-cap: "Available samples from 25 participants in the sample data set.
#| RNA-seq was performed on muscle biopsy samples obtained prior to (w0),
#| after two (w2pre, and w2post), and twelve weeks of resistance training
#| perfomed with multiple and single sets allocated to either leg (condition)."
#| echo: false
#| fig-width: 6
#| fig-height: 8


rna_seq_metadata %>%
  filter(rnaseq_include == "incl") %>%
  mutate(time = factor(time, levels = c("w0", "w2pre", "w2post", "w12"))) %>%
  ggplot(aes(participant, time,
    fill = condition,
    group = paste(condition, participant)
  )) +
  geom_line(position = position_dodge(width = 0.7)) +
  geom_point(
    position = position_dodge(width = 0.7),
    shape = 21
  ) +
  coord_flip()
```


The structure of the data set allows for a mixed-effects negative binomial model

$$
\begin{align*}
Y_{i} & \sim \text{NB}(\mu_{i}, \phi) \\
\operatorname{log}(\mu_{i}) & = X_{i}\beta + Z_{i}b
\end{align*}
$$

where $Y_{i}$ are the counts for observation $i$, modelled using the negative binomial distribution with and average $\mu$ and dispersion $\phi$ using a log-link. Design variables, time and condition, are included in $X$. The design matrix ($X_{i}$) allows for an interaction between time and condition and the design matrix for the random effects ($Z_i$) allows for a random intercept per participant.

Since we are using a negative binomial model we need to provide integers as input. We will round and convert to integers  


Model parameters are estimated for all genes in the data set (`rna_seq_sample`). Model fitting can be performed with parallel computing using multiple cores. First we detect the number of available cores and initiate the process using all available cores. We will use only a subset of transcripts in this demonstration, we can use the subset to make an index to subset the number of targets.

```{r}
# Change all counts to integer
rna_seq_sample_integer <- rna_seq_sample %>%
  dplyr::mutate(across(-transcript_id, ~ as.integer(round(.x, 0)))) %>%
  print(n = 5)

# Remove missing values from the meta data
rna_seq_metadata <- rna_seq_metadata %>%
  filter(!is.na(seq_sample_id))



## Check the number of cores available
ncores <- parallel::detectCores()

## Save a list of arguments used by the fitting algorithm
args <- list(
  formula = y ~ time * condition + (1 | participant),
  family = glmmTMB::nbinom2
)


fits <- seqwrap(
  fitting_fun = glmmTMB::glmmTMB,
  arguments = args,
  data = rna_seq_sample_integer,
  metadata = rna_seq_metadata,
  samplename = "seq_sample_id",
  summary_fun = NULL,
  eval_fun = NULL,
  additional_vars = NULL,
  exported = list(),
  subset = 1:10,
  cores = ncores
)
```

Since we did not provide a function for `summary_fun` or `eval_fun` the function returns a list of fitted models. `seqwrap` accepts user-defined summary functions. To build such a function we can use a single model from our list of fitted models. Note that models are stored under `models` in the list provided by `seqwrap`.


```{r}
# A single model is stored for testing our function
x <- fits$models[[1]]




sum_fun <- function(x) {
  cond_effects <- data.frame(
    cbind(
      data.frame(
        coef = rownames(coef(summary(x))$cond)
      )
    ),
    coef(summary(x))$cond,
    row.names = NULL
  )

  return(cond_effects)
}


sum_fun(x)
```

The provided function gives us a data frame of coefficients from the fitted model. `seqwrap` will create a list of data frames each created with the summary function. For illustrative purposes we will refit models with our summary function. 

```{r}
fits2 <- seqwrap(
  fitting_fun = glmmTMB::glmmTMB,
  arguments = args,
  data = rna_seq_sample_integer,
  metadata = rna_seq_metadata,
  samplename = "seq_sample_id",
  summary_fun = sum_fun, ## Our function
  eval_fun = NULL,
  additional_vars = NULL,
  exported = list(),
  subset = 1:100,
  cores = ncores
)


# The summarized models are stored as a named list
names(fits2$summaries)
```

Note that the summarized models are a named list, corresponding to the list of model objects. To combine the transcript name with the corresponding table of coefficients we can simply replicate the name from the list when combining the data using `bind_rows`. The resulting combined data can be plotted, e.g. as in @fig-vulcano 

```{r}
#| label: fig-vulcano


bind_rows(fits2$summaries) %>%
  mutate(target = rep(names(fits2$summaries), each = 8)) %>%
  filter(coef != "(Intercept)") %>%
  ggplot(aes(Estimate, -log(Pr...z..))) +
  geom_point() +
  facet_wrap(~coef)
```

Note that some estimates are quite large (x-axis in @fig-vulcano). This might indicate some difficulties in fitting the model. We might want to make sure the models meet assumptions of the negative binomial model. 

We can add another user specified function that evaluates models. The [`DHARMa` package](https://cran.r-project.org/web/packages/DHARMa/vignettes/DHARMa.html) can help to perform residual diagnostics for generalized linear mixed models using simulations. 

The function constructed below takes a model (`x`) and performs simulations to construct residuals. Each model is given a set of 


```{r}
library(DHARMa)

eval_mod <- function(x) {
  sim <- DHARMa::simulateResiduals(x, n = 1000)
  unif <- DHARMa::testUniformity(sim, plot = FALSE)
  zinfl <- DHARMa::testZeroInflation(sim, plot = FALSE)
  results <- data.frame(
    pval.unif = unif$p.value,
    pval.zinfl = zinfl$p.value
  )

  return(results)
}
```

For illustrative purposes we re-fit everything and add the evaluation function to `seqwrap`.


```{r}
fits3 <- seqwrap(
  fitting_fun = glmmTMB::glmmTMB,
  arguments = args,
  data = rna_seq_sample_integer,
  metadata = rna_seq_metadata,
  samplename = "seq_sample_id",
  summary_fun = sum_fun, ## Our function
  eval_fun = eval_mod,
  additional_vars = NULL,
  exported = list(),
  subset = 1:100,
  cores = ncores
)
```


We can now combine the information provided from the summary and evaluation function.


```{r}
bind_rows(fits3$summaries) %>%
  mutate(target = rep(names(fits3$summaries), each = 8)) %>%
  inner_join(bind_rows(fits3$evaluations) %>%
    mutate(target = names(fits3$evaluations))) %>%
  filter(
    coef != "(Intercept)",
    abs(Estimate) > 10
  ) %>%
  ggplot(aes(Estimate, -log(Pr...z..))) +
  geom_point(aes(color = pval.unif)) +
  facet_wrap(~coef)


fits3$model_fits$`ENST00000629380.3_SYNGAP1-208`

rna_seq_sample_integer %>%
  filter(transcript_id == "ENST00000629380.3_SYNGAP1-208")
```




## References
